File: SethMiller-TypeScript/src/formatter.ts
import { ParsedObject } from "./common";

const INDENT_SIZE = 2;

export const formatObjectAsString = (obj: ParsedObject, shouldSort: boolean = false, indentationLevel: number = 0): string => {
    let result = '';
    const indent = ' '.repeat(INDENT_SIZE * indentationLevel);

    for (let k of (shouldSort ? Object.keys(obj).sort() : Object.keys(obj))) {
        const nestedObjectResult = typeof (obj[k]) === 'object' ?
            formatObjectAsString(obj[k] as ParsedObject, shouldSort, indentationLevel + 1) :
            '';
        result += `${indent}- ${k}\n${nestedObjectResult}`;
    }

    return result;
}

File: SethMiller-CSharp/Source/Logic/Formatter.cs
namespace SethMiller_CSharp.Logic
{
    public static class Formatter
    {
        public static readonly int IndentSize = 2;

        public static string FormatObjectAsString(ParsedObject obj, bool shouldSort = false, int indentationLevel = 0)
        {
            var result = "";
            var indent = new string(' ', indentationLevel * IndentSize);
            var keys = shouldSort ? obj.Keys.Order().AsEnumerable() : obj.Keys;

            foreach (var k in keys)
            {
                var nestedObjectResult = obj[k] != null ?
                    FormatObjectAsString(obj[k]!, shouldSort, indentationLevel + 1) 
                    : "";
                result += $"{indent}- {k}{Environment.NewLine}{nestedObjectResult}";
            }

            return result;
        }
    }
}
